File: CSharpCodeBase/utils/ailib/ainode.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ainode {
 public void init (self, entity, childNodes){
       this.childNodes = childNodes;
       this.entity = entity;
       this.status = NODE_READY;
       this.lastStatus = NODE_READY;
       this.parent = null;
       if(this.childNodes != null  ){ 
         for(k, v in pairs(this.childNodes) ){
           v.parent = self;
         }
       }
       this._lockstart = 0;
       this._locktime = 0;
 }})
 public void Visit(){
   this.status = NODE_FAILURE;
 }
 public void Save(){
   this.lastStatus = this.status;
   if(this.childNodes != null  ){ 
     for(k, v in pairs(this.childNodes) ){
       v:Save();
     }
   }
 }
 public void Process(){
   if(this.status != NODE_RUNNING  ){ 
     self:Reset();
   }else{if this.childNodes != null  ){ 
     for(k, v in pairs(this.childNodes) ){
       v:Process();
     }
   }
 }
 public void Reset(){
   if(this.status != NODE_READY  ){ 
     this.status = NODE_READY;
     if(this.childNodes != null  ){ 
       for(k, v in pairs(this.childNodes) ){
         v:Reset();
       }
     }
   }
 }
 public void Sleep(seconds){
   this._lockstart = os.clock();
   this._locktime = seconds;
 }
 public void IsSleeping(){
   return os.clock() - this._lockstart < this._locktime;
 }
 public void OnEvent(data){
   if(this.childNodes != null  ){ 
     for(k, v in pairs(this.childNodes) ){
       if(v.status == NODE_RUNNING  ){ 
         v:OnEvent(data);
       }
     }
   }  
 }
 }}

File: CSharpCodeBase/utils/ailib/ainode.lua
local Class = require 'utils.hump.class'
local AITree = require 'utils.ailib.aitree'

local AINode = Class({init = function (self, entity, childNodes)
      self.childNodes = childNodes
      self.entity = entity
      self.status = NODE_READY
      self.lastStatus = NODE_READY
      self.parent = nil
      if self.childNodes ~= nil then
        for k, v in pairs(self.childNodes) do
          v.parent = self
        end
      end
      self._lockstart = 0
      self._locktime = 0
end})

function AINode:Visit()
  self.status = NODE_FAILURE
end

function AINode:Save()
  self.lastStatus = self.status
  if self.childNodes ~= nil then
    for k, v in pairs(self.childNodes) do
      v:Save()
    end
  end
end

function AINode:Process()
  if self.status ~= NODE_RUNNING then
    self:Reset()
  elseif self.childNodes ~= nil then
    for k, v in pairs(self.childNodes) do
      v:Process()
    end
  end
end

function AINode:Reset()
  if self.status ~= NODE_READY then
    self.status = NODE_READY
    if self.childNodes ~= nil then
      for k, v in pairs(self.childNodes) do
        v:Reset()
      end
    end
  end
end

function AINode:Sleep(seconds)
  self._lockstart = os.clock()
  self._locktime = seconds
end

function AINode:IsSleeping()
  return os.clock() - self._lockstart < self._locktime
end

function AINode:OnEvent(data)
  if self.childNodes ~= nil then
    for k, v in pairs(self.childNodes) do
      if v.status == NODE_RUNNING then
        v:OnEvent(data)
      end
    end
  end  
end

return AINode